client/auth/credentials.go

```diff
@@ -0,0 +1,125 @@
+package auth
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"errors"
+	"fmt"
+	"net/http"
+)
+
+var (
+	ErrIncorrectCredentials = errors.New("incorrect credentials")
+	ErrChallenged           = errors.New("user password needs to be reset")
+	ErrTooManyRequests      = errors.New("too many requests to Enlight SSO")
+	ErrInactivated          = errors.New("user has been inactivated")
+	ErrUnknownTokenType     = errors.New("provided token type not present in response")
+)
+
+const (
+	DefaultTokenType = "identityToken"
+)
+
+type CredentialsTokenProvider struct {
+	Username string
+	Password string
+	Endpoint string
+
+	Client    CredentialsClient
+	TokenType string
+}
+
+type CredentialsClient interface {
+	Do(*http.Request) (*http.Response, error)
+}
+
+type SignInRequest struct {
+	Username string `json:"username"`
+	Password string `json:"password"`
+}
+
+type SignInResponse struct {
+	Tokens    map[string]RawToken `json:"tokens"`
+	Challenge *struct {
+		ID   string `json:"id"`
+		Type string `json:"type"`
+	} `json:"challenge"`
+}
+
+func (provider *CredentialsTokenProvider) GetRawToken(ctx context.Context) (RawToken, error) {
+	if provider.TokenType == "" {
+		provider.TokenType = DefaultTokenType
+	}
```

Comment:
In general, I try to avoid the constructor pattern. In some places where you need to initialize private members, it is unavoidable. One reason for the choice I made here was to avoid all the boilerplate of adding a bunch of setters or a configuration struct.